src/Cli/dotnet/commands/dotnet-test/SolutionAndProjectUtility.cs

```diff
@@ -102,51 +102,88 @@ public static string GetRootDirectory(string solutionOrProjectFilePath)
 
         public static IEnumerable<Module> GetProjectProperties(string projectFilePath, IDictionary<string, string> globalProperties, ProjectCollection projectCollection)
         {
-            var project = projectCollection.LoadProject(projectFilePath, globalProperties, null);
-            return GetModulesFromProject(project);
+            var projects = new List<Module>();
+
+            var globalPropertiesWithoutTargetFramework = new Dictionary<string, string>(globalProperties);
+            globalPropertiesWithoutTargetFramework.Remove(ProjectProperties.TargetFramework);
+
+            var project = projectCollection.LoadProject(projectFilePath, globalPropertiesWithoutTargetFramework, null);
+
+            // Check if TargetFramework is specified in global properties
+            if (globalProperties.TryGetValue(ProjectProperties.TargetFramework, out string targetFramework))
+            {
+                if (IsValidTargetFramework(project, targetFramework))
+                {
+                    project.SetProperty(ProjectProperties.TargetFramework, targetFramework);
+                    project.ReevaluateIfNecessary();
+                    var module = GetModuleFromProject(project);
+                    if (module != null)
+                    {
+                        projects.Add(module);
+                    }
+                }
+            }
+            else
+            {
+                string targetFrameworks = project.GetPropertyValue(ProjectProperties.TargetFrameworks);
+
+                if (string.IsNullOrEmpty(targetFrameworks))
+                {
+                    var module = GetModuleFromProject(project);
+                    if (module != null)
+                    {
+                        projects.Add(module);
+                    }
+                }
+                else
+                {
+                    var frameworks = targetFrameworks.Split(CliConstants.SemiColon, StringSplitOptions.RemoveEmptyEntries);
+                    foreach (var framework in frameworks)
+                    {
+                        project.SetProperty(ProjectProperties.TargetFramework, framework);
+                        project.ReevaluateIfNecessary();
+                        var module = GetModuleFromProject(project);
+                        if (module != null)
```

Comment:
Nit: pattern matching

```suggestion
                        if (GetModuleFromProject(project) is {} module)
```